src/main/java/com/epam/bookstoreservice/service/impl/BookstoreServiceImpl.java

```diff
@@ -144,29 +117,95 @@ public List<BookResponseDTO> getBooksByCategoryAndKeyWord(String category, Strin
+        Optional<BookEntity> bookFindById = bookDao.findById(id);
+
+        return bookFindById.orElseThrow(BookNotFoundException::new);
+    }
+
+    private void checkInventoryOfABook(BookEntity bookEntity, Integer salesNumber) {
+
+        int remainingInventory = bookEntity.getTotalCount() - salesNumber;
+
+        if (remainingInventory < 0) {
+            throw new InsufficientInventoryException();
         }
-        List<BookEntity> bookEntityListFindByCategoryAndKeyword = bookDao.findByCategoryAndKeyword(category, bookId, keyword);
 
-        if (bookEntityListFindByCategoryAndKeyword.isEmpty()) {
-            throw new BookNotFoundException();
+    }
+
+    private void checkInventoryOfBooks(List<SellDTO> sellDTOList) {
+        sellDTOList.forEach(sellDTO ->
+                checkInventoryOfABook(checkAndGetABookEntityFindByIdIfExist(sellDTO.getId()), sellDTO.getSalesNumber()));
+    }
+
+    private BookResponseDTO sellBookByIdAndSalesNumber(BookEntity bookEntity, Integer salesNumber) {
+
+        bookEntity.setSold(bookEntity.getSold() + salesNumber);
+        bookEntity.setTotalCount(bookEntity.getTotalCount() - salesNumber);
+
+        return bookDtoAndBookEntityMapper.entityToResponseDto(bookDao.save(bookEntity));
+    }
+
+    private void validId(Integer id, BookRequestDTO bookRequestDTO) {
+
+        boolean idNonNullAndUnmatched
+                = Objects.nonNull(bookRequestDTO.getId()) && !Objects.equals(id, bookRequestDTO.getId());
+
+        if (idNonNullAndUnmatched) {
+            throw new UnmatchedIdException();
         }
+    }
+
+    private List<BookResponseDTO> getBooksByCategoryAndKeyWordUtil(String category, String keyword) {
+
+        List<BookEntity> bookEntityListFindByCategoryAndKeyword = bookDao
+                .findByCategoryAndKeyword(category, getBookIdWhenKeywordIsNumber(keyword), keyword);
+
+        checkBookEntityListIfExist(bookEntityListFindByCategoryAndKeyword);
 
         return bookEntityListFindByCategoryAndKeyword
                 .stream()
-                .map(bookRequestDtoToBookEntity::entityToResponseDto).collect(Collectors.toList());
+                .map(bookDtoAndBookEntityMapper::entityToResponseDto).collect(Collectors.toList());
     }
+
+    private void checkBookEntityListIfExist(List<BookEntity> bookEntityListFindByCategoryAndKeyword) {
+        if (bookEntityListFindByCategoryAndKeyword.isEmpty()) {
+            throw new BookNotFoundException();
+        }
+    }
+
+    private int getBookIdWhenKeywordIsNumber(String keyword) {
+        if (checkKeyWordIsDigit(keyword)){
+            return Integer.parseInt(keyword);
+        }
+        return DEFAULT_BOOK_ID;
+    }
+
+    private boolean checkKeyWordIsDigit(String keyword) {
+        char[] keywordChars = keyword.toCharArray();
+
+        for (char keywordChar : keywordChars) {
+            boolean keywordCharIsNotDigit = !Character.isDigit(keywordChar);
```

Comment:
It's better to use isNotXXX than !isXXX.
And if it has only one condition  you can just use it like 
if (isNotDigit(xxx)){
   ...
}